lib/parseHmrOption.js

```diff
@@ -0,0 +1,36 @@
+// Detects the life cycle of npm for determination of whether it's production distribution
+var NPM_PREPUBLISH_EVENT_REGEX = /^prepublish(Only)?$/gi;
+
+/**
+ * Parse `options.hmr` value depends on the following condition:
+ *   1. If `options.hmr` is a set value, return itself
+ *   2. If `options.hmr` is missing with value `undefined`, try to parse the npm command used to
+ *     invoke webpack via npm lifecycle event, to determine the default value:
+ *       a) If invoked via npm `prepublish` or `prepublishOnly`, this is production distribution and
+ *         default value for `options.hmr` to be false.
+ *       b) If invoked via other npm scripts or non-npm commands, assuming this is local or dev env,
+ *         default value for `options.hmr` to be true.
+ * @param {Boolean|undefined} hmr - The original `options.hmr`
+ * @returns {Boolean}
+ */
+module.exports = function(hmr) {
```

Comment:
Does this work even if the `package.json` is like this:

```json
{
  "scripts": {
    "build": "webpack",
    "prepublishOnly": "npm run build"
  }
}
```
?

Reply:
Hmmm, npm lifecycle event seems only picking up the immediate one, so in this case, it's `build`. I'll need to think about a better solution here.